### PR TITLE
User flavored compression

### PR DESCRIPTION
**Pull request by patrickn2**
In this PR I created a way to the user chose what he is going to use to compress the data. he could use zlib, gzip, etc...
I think this is a better version.

### CODE REVIEW

File: config.go [hunk 1]
```diff
@@ -53,5 +61,9 @@ func DefaultConfig() *Config {
 		TcpKeepAlive:       true,
 		TcpKeepAlivePeriod: 60 * time.Second,
 		TcpNoDelay:         true,
+		Compression: struct {
+			unzip   func(value string) (string, error)
+			deflate func(value string) (string, error)
```
[the remark this answers is not in the document]
Comment (dterei) [reply config.go:66]: And wouldn't `[]byte` rather than `string` be better choice?

File: client.go [hunk 2]
```diff
@@ -157,6 +157,12 @@ func (c *Client) getCAS(key string, ocas uint64) (val string, flags uint32, cas
 	}
 
 	err = c.perform(m)
+	if c.config.Compression.unzip != nil && err == nil {
```
Comment (dterei) [inline client.go:160]: Probably best to check `config.Compression != nil` too for being as safe as possible

**Comment by dterei** [comment]
Thanks @patrickn2! Left a review. Be good to provide support for using say zlib and gzip compression out of the box too.

File: utils_test.go [hunk 3]
```diff
@@ -1,10 +1,53 @@
...
+		return compressedValue.String(), nil
+	}
+
+	config.Compression.unzip = func(value string) (string, error) {
+		if value == "" {
+			return value, nil
+		}
+		zr, err := zlib.NewReader(strings.NewReader(value))
+		if err != nil {
+			return value, nil // Does not return error, the value could be not compressed
```
[the remark this answers is not in the document]
Comment (patrickn2) [reply utils_test.go:34]: Backwards compatibility, if the old data is not compressed we should return the data anyway

[the remark this answers is not in the document]
Comment (dterei) [reply utils_test.go:34]: Do we need to be backwards compatible in test code though?

File: client.go (hunk 2 above)
Comment (patrickn2) [reply client.go:160]: ok, done

File: client.go [hunk 4]
```diff
@@ -157,6 +157,12 @@ func (c *Client) getCAS(key string, ocas uint64) (val string, flags uint32, cas
...
+		m.val, err = c.config.Compression.unzip(m.val)
+		if err != nil {
```
[the remark this answers is not in the document]
Comment (patrickn2) [reply client.go:162]: Sure we can

File: config.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (patrickn2) [reply config.go:66]: decompress and compress ok

Now about the []byte, the ideia is to use the same datatype that set and get uses

File: client.go [hunk 5]
```diff
@@ -231,7 +237,12 @@ func (c *Client) setGeneric(op opCode, key, val string, ocas uint64, flags, exp
 		key:     key,
 		val:     val,
 	}
-
+	if c.config.Compression.deflate != nil {
```
[the remark this answers is not in the document]
Comment (patrickn2) [reply client.go:240]: I tried but the compiler gives me a message "invalid operation: c.config.Compression != nil (mismatched types struct{decompress func(value string) (string, error); compress func(value string) (string, error)} and untyped nil)compilerMismatchedTypes"

I created a snippet of the same code to check if this is going to throw an error and did not.



```
	package main

import (
	"fmt"
)

type Config struct {
	Compression struct {
		decompress func(value string) (string, error)
		compress   func(value string) (string, error)
	}
}

func printConfig(c *Config) {
	if c.Compression.compress != nil {
		fmt.Println(c.Compression.compress("asasas"))
	}
}


func main() {
	aa := &Config{}
	printConfig(aa)
}

File: utils_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (patrickn2) [reply utils_test.go:34]: Oh, my mistake, I was thinking about another matter :D

File: utils_test.go [hunk 6]
```diff
@@ -1,10 +1,53 @@
...
+		}
+		defer zr.Close()
+		var unCompressedValue bytes.Buffer
+		_, err = io.Copy(&unCompressedValue, zr)
+		if err != nil {
+			return value, nil
```
[the remark this answers is not in the document]
Comment (patrickn2) [reply utils_test.go]: Solved

File: config.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (dterei) [reply config.go]: I see, fair enough then

**Review by dterei — APPROVED** [review]

**Comment by patrickn2** [comment]
> Thanks @patrickn2! Left a review. Be good to provide support for using say zlib and gzip compression out of the box too.

if you want I can add this to the Readme.md but their behavior are just the same, you only have to replace the import and the lib name.

**Comment by dterei** [comment]
Yeah that would be good, and sorry, i merged the diff so also include any changes requested that weren't merged please

**Comment by patrickn2** [comment]
> Yeah that would be good, and sorry, i merged the diff so also include any changes requested that weren't merged please

Don't Worry
I'm gonna create another branch and PR

**Pull request merged**